botocore/client.py

```diff
@@ -325,6 +246,144 @@ def _api_call(self, *args, **kwargs):
         return _api_call
 
 
+class ClientEndpointBridge(object):
+    """Bridges endpoint data and client creation
+
+    This class handles taking out the relevant arguments from the endpoint
+    resolver and determining which values to use, taking into account any
+    client configuration options and scope configuration options.
+
+    This class also handles determining what, if any, region to use if no
+    explicit region setting is provided. For example, Amazon S3 client will
+    utilize "us-east-1" by default if no region can be resolved."""
+
+    SIGNATURE_PRECEDENCE = ['s3', 's3v4', 'v4', 'v2']
```

Comment:
Are there any specific services that use the "prefer v4 over v2" logic now?  I'm wondering if the s3->s3v4 is a special case because we can't migrate to s3v4 in a backwards compatible way just yet.  The prefer v4 over v2 case would only be triggered if a service endpoint has `['v4', 'v2']` in its `signatureVersion` list right?  Given that there's no entries right now that have that, and future services/endpoints will likely use v4, is this necessary?

Just trying to understand if this is addressing an existing issue we have, is more of a future proof thing, etc.